apache/superset
# fix: SSH Tunnel configuration settings

### SUMMARY
This PR fixes several issues with the SSH Tunnel option in the Database Modal which are related to edit the SSH Tunnel config, enable/disable the option, and misc fixes/enhancements. In particular:

- Removes a duplicate `SSHTunnelSwitch` component from dynamic forms and merges the logics of both dynamic and SQLAlchemy forms into one overridable component
- Fixes an issue that made removing the SSH tunnel config impossible when editing a database connection
- Delegates the deletion of a ssh tunnel to the `UpdateDatabaseCommand`
- Fixes an issue with the SSH tunnel config still being included in the request even when toggled off
- Fixes an issue for which the switch was showing also with `SSH_TUNNELING` OFF.  Saving would cause a "SSH Tunneling is not enabled" error
- Surfaces to the user a missing port error in the SQLAlchemy uri when enabling SSH Tunnel (the port is required for the tunnel to work)
- Clean the SSH Tunnel config from unnecessary credentials when switching from password to private key
- Improves the UX so that toggling ON and OFF the SSH Tunnel config won't clear the form until saved
- Misc fixes

### BEFORE

https://github.com/apache/superset/assets/60598000/c38052d6-53b0-47d8-a87a-391a647085d9

### AFTER

https://github.com/apache/superset/assets/60598000/f65ad7da-962f-4500-82f2-eb42bd2139ba

### TESTING INSTRUCTIONS
1. Create a new database connection using a SQLAlchemy uri and dynamic forms, make sure SSH Tunnel can be toggled ON and OFF and that configs can be saved successfully
2. Edit a database connection that has a SSH Tunnel config, make sure the SSH Tunnel can be toggled ON and OFF and that it keeps the config available to the user until toggled OFF and saved
3. Edit a database connection that has a SSH Tunnel config, make sure any change to the SSH tunnel can be saved successfully
4. Re-open the database config and make sure the SSH Tunnel has been deleted successfully
5. Create or edit a database connection using a SQLAlchemy uri without specifying the port, make sure an error is surfaced to the UI informing that specifying a port is necessary

### ADDITIONAL INFORMATION
